### PR TITLE
Replace public config store with RPC methods/notifications

### PR DESCRIPTION
**Pull request by rekmarks**
- Deletes the public config stream, `_publicConfigStore`, and `publicConfigStore` getter
- Gets state previously received through public config stream with RPC method and notifications
  - New RPC Methods
    - `wallet_getProviderState`
  - New RPC Notifications
    - `wallet_unlockStateChanged`
    - `wallet_chainChanged`

Corresponding PR in extension: https://github.com/MetaMask/metamask-extension/pull/8640

CI: https://app.circleci.com/pipelines/github/MetaMask/inpage-provider?branch=delete-public-config

### CODE REVIEW

**Comment by Gudahtt** [comment]
How does this relate to the breaking changes? Did you just want to bundle it with them for the sake of avoiding another breaking change release, or is it functionally tied to them in some way?

**Comment by rekmarks** [comment]
Ah, we're bundling this breaking change with all the rest just for the sake of convenience!

**Review by Gudahtt — COMMENTED** [review]
Should we be using `metamask_` here instead of `wallet_`? :thinking: 

I had thought we were, as a general rule, using `metamask_` for methods that are internal-only or that we have no plans to standardize, and `wallet_` for things that _might_ be standardized at some point. These methods seem to me of internal relevance only, at least at the moment.

File: src/MetaMaskInpageProvider.js [hunk 1]
```diff
@@ -356,16 +302,50 @@ module.exports = class MetaMaskInpageProvider extends SafeEventEmitter {
...
+   * 'eth_accounts', and emits necessary events.
+   *
+   * @private
+   */
+  _initializeState () {
+    Promise.all([
+      this.request({ method: 'wallet_getProviderState' }),
+      this.request({ method: 'eth_accounts' }),
+    ])
+      .then(([state, accounts]) => {
```
Comment (Gudahtt) [inline src/MetaMaskInpageProvider.js]: Nit: `async / await` should be preferred over `.then / .catch`

**Comment by rekmarks** [comment]
@Gudahtt, good point. I had considered that but never executed. Done in: 2fe974d

File: src/siteMetadata.js [hunk 2]
```diff
@@ -18,7 +18,7 @@ async function sendSiteMetadata (engine, log) {
     // call engine.handle directly to avoid normal RPC request handling
     engine.handle(
       {
-        method: 'wallet_sendDomainMetadata',
+        method: 'metamask_sendDomainMetadata',
```
Comment (Gudahtt) [inline src/siteMetadata.js:21]: This seems out of scope for this PR, and `wallet_` might be the more appropriate prefix for this method anyway. We had talked about encouraging dapps to set their own metadata. So we might want to consider documenting this as-is.

File: src/MetaMaskInpageProvider.js [hunk 3]
```diff
@@ -485,19 +537,13 @@ module.exports = class MetaMaskInpageProvider extends SafeEventEmitter {
          * @returns {Promise<boolean>} - Promise resolving to true if MetaMask is currently unlocked
          */
         isUnlocked: async () => {
-          if (this._state.isUnlocked === undefined) {
```
Comment (Gudahtt) [inline src/MetaMaskInpageProvider.js:488]: This is an unfortunate change in behaviour. This used to block on determining the actual unlock status of the wallet, but now it returns `false` right away if called during initialization, even if the wallet is unlocked.

File: src/MetaMaskInpageProvider.js [hunk 4]
```diff
@@ -459,8 +441,77 @@ module.exports = class MetaMaskInpageProvider extends SafeEventEmitter {
...
+
+    if (isUnlocked !== this._state.isUnlocked) {
+      this._state.isUnlocked = isUnlocked
+
+      if (isUnlocked) {
+
+        // this will get the exposed accounts, if any
+        try {
+          this._rpcRequest(
+            { method: 'eth_accounts', params: [] },
```
Comment (Gudahtt) [inline src/MetaMaskInpageProvider.js:503]: Could we bundle the accounts with the unlock notification, to avoid this second call? It would simplify things a great deal. I think it would let us remove the entire `isInternal` parameter altogether, and that `'eth_accounts' unexpectedly updated accounts` warning

Comment (Gudahtt) [reply src/MetaMaskInpageProvider.js:503]: I guess this can wait for a subsequent PR either way though.

File: src/MetaMaskInpageProvider.js [hunk 5]
```diff
@@ -459,8 +441,77 @@ module.exports = class MetaMaskInpageProvider extends SafeEventEmitter {
...
+      )
+      return
+    }
+
+    if (chainId !== this.chainId || networkVersion !== this.networkVersion) {
+      this.chainId = chainId
+      this.emit('chainChanged', this.chainId)
+
+      this.networkVersion = networkVersion
+      this.emit('networkChanged', this.networkVersion)
```
Comment (Gudahtt) [inline src/MetaMaskInpageProvider.js:477]: This would re-emit `networkChanged` even if it remained the same (e.g. if only the `chainId` changed). Was this intentional? It is a change in behaviour from before.

File: src/MetaMaskInpageProvider.js [hunk 6]
```diff
@@ -356,16 +302,50 @@ module.exports = class MetaMaskInpageProvider extends SafeEventEmitter {
...
+   * Constructor helper.
+   * Populates initial state by calling 'metamask_getProviderState' and
+   * 'eth_accounts', and emits necessary events.
+   *
+   * @private
+   */
+  _initializeState () {
+    Promise.all([
+      this.request({ method: 'metamask_getProviderState' }),
+      this.request({ method: 'eth_accounts' }),
```
Comment (Gudahtt) [inline src/MetaMaskInpageProvider.js]: Any reason `eth_accounts` is a separate call, rather than being included with the provider state?

Comment (Gudahtt) [reply src/MetaMaskInpageProvider.js]: Not that I object to us doing this for now - we can optimize this later if we feel like it. Just wondering if there was a reason beyond that this is closer to the old behaviour.

File: src/MetaMaskInpageProvider.js (hunk 5 above)
Comment (rekmarks) [reply src/MetaMaskInpageProvider.js:477]: Hm, the only case where the chain ID but _not_ the network ID changes is for the few cases where chains share the latter but not the former. "changes" here we'll take to mean "becomes a different value in state."

In those cases, I think it's fair to say that "the network changed" even though the network ID didn't. I'm really agnostic here, since we consider the the network ID and its related event deprecated.

File: src/MetaMaskInpageProvider.js (hunk 4 above)
Comment (rekmarks) [reply src/MetaMaskInpageProvider.js:503]: Hm, this is actually very feasible now. Let's do that in a follow-up!

File: src/siteMetadata.js (hunk 2 above)
Comment (rekmarks) [reply src/siteMetadata.js:21]: I always considered it to be kind of a hack, personally, and I think there's more work to be done on specifying the method before we commit to publishing its interface.

As to the out of scope, yeah kind of, but while we're here? 😅

File: src/MetaMaskInpageProvider.js (hunk 1 above)
Comment (rekmarks) [reply src/MetaMaskInpageProvider.js]: Done here: 79128b4

File: src/MetaMaskInpageProvider.js (hunk 6 above)
Comment (rekmarks) [reply src/MetaMaskInpageProvider.js]: Done here: 79128b4

The `getProviderState` implementation was updated on the extension side as well. I'll thoroughly test that with this version of the inpage provider before opening the extension PR.

File: src/MetaMaskInpageProvider.js (hunk 3 above)
Comment (rekmarks) [reply src/MetaMaskInpageProvider.js:488]: I restored that behavior here: c0b32ae

File: src/MetaMaskInpageProvider.js (hunk 5 above)
Comment (rekmarks) [reply src/MetaMaskInpageProvider.js:477]: EIP-1193 doesn't specify what to do, so it's really a judgment call: https://eips.ethereum.org/EIPS/eip-1193#networkchanged-deprecated

I'd say that preserving existing behavior is best, but the network _has actually changed_ whenever the chain ID value but not the network ID value changes.

Comment (Gudahtt) [reply src/MetaMaskInpageProvider.js:477]: Yeah. If not for legacy reasons, I'd agree that re-emitting makes the most sense. But since this behaviour might be relied upon, and since we're deprecating it anyway, I think it'd be better to preserve the old behaviour.

File: src/siteMetadata.js (hunk 2 above)
Comment (Gudahtt) [reply src/siteMetadata.js:21]: >I think there's more work to be done on specifying the method before we commit to publishing its interface.

Makes sense. I hadn't really thought through whether it was ready to be public-facing in its current form.

It would have been better if this was `metamask_` in the first place., but... it still seems better to leave it as is for now :man_shrugging: Technically dapps might already be sending this? Also the less provider changes we make, the less broken things with embedded providers will be (e.g. extensions, and dapps that import our provider as a workaround to that Firefox CSP bug). But I guess I won't block on it. Embedded providers will be _fairly broken_ either way.

File: src/MetaMaskInpageProvider.js (hunk 5 above)
Comment (rekmarks) [reply src/MetaMaskInpageProvider.js:477]: Aight, I restored the behavior here: 94d84d6

File: src/siteMetadata.js (hunk 2 above)
Comment (Gudahtt) [reply src/siteMetadata.js:21]: Actually... maybe we can preserve the public config store in the extension to support old embedded providers, at least for a short while? :thinking: I don't know how feasible this is - I'll have to look more closely at the extension PRs.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Comment by rekmarks** [comment]
CI passes: https://app.circleci.com/pipelines/github/MetaMask/inpage-provider/388/workflows/e6397b45-6ba6-4d5b-ae4d-f518d981cb28
